### ISSUE TITLE
[bug] Memory leak when using `faidx::Reader::fetch_seq()` ?

### ISSUE THREAD

**Issue by MaelLefeuvre** [rust-bio/rust-htslib#401]
# Bug description

Hello,

I'm currently working on a simple rust utility that requires me to read and compare bam records with its matching reference sequence. 

My approach ended up using the `rust_htslib::faidx::Reader::fetch_seq()`. However, i'm now noticing my binary is now experiencing quite the serious memory leak (see screen captures below)

Profiling the binary through [`heaptrack`](https://github.com/KDE/heaptrack), clearly points out to the ffi call to `htslib::faidx_fetch_seq64()` within the `faidx::Reader::fetch_seq()` method (See [`src/faidx/mod.rs:73-79`](https://github.com/rust-bio/rust-htslib/blob/e91b1e2a5ca98d085a02343207100a3aea7a0aec/src/faidx/mod.rs#L73-L79))

| Timeline | Summary | Bottom-up stacktrace |
| -------- | ------- | -------------------- |
| ![pmd-mask-memory-consumption-timeline](https://github.com/MaelLefeuvre/pmd-mask/assets/70585821/20b440ac-da12-4343-b071-7136871d6bc0) | ![image](https://github.com/MaelLefeuvre/pmd-mask/assets/70585821/ff43b6fa-e549-474f-b4e8-fd419a2253e0) | ![image](https://github.com/MaelLefeuvre/pmd-mask/assets/70585821/2b2c3720-6c17-4182-b8cc-a32a9f559292) |

# Minimally reproducible example

A simple loop such as the one below is enough to create a steady increase of the binary's resident set size, while I would expect the memory usage to pretty much constant in this example. 

```
use rust_htslib::faidx;
use rand::{thread_rng, Rng};

fn main() {
    let fasta     = std::env::args().nth(1).expect("No reference provided");
    let reference = faidx::Reader::from_path(&fasta).unwrap();

    let mut rng = thread_rng();
    loop {
        // ---- Fetch a random chromosome, position and read length within a reference genome 
        let chr   = rng.gen_range(1..=22u32).to_string();
        let len   = rng.gen_range(10..50usize);
        let start = rng.gen_range(0..100usize - len);
        let refseq = reference.fetch_seq(chr, start, start + len).unwrap();
        println!("{}", unsafe{ std::str::from_utf8_unchecked(refseq)});
    }
}
```

A small public github repository containing the example above, along with a dummy reference file can be found here: [rust-htslib-memleak-mre](https://github.com/MaelLefeuvre/rust-htslib-memleak-mre)

# Some thoughts:

Of course, I might be completely misusing the `fetch_seq()` method - hence the '?' in this issue's title. Regardless, be it a safety or documentation issue, I'm thinking this might need some resolving.

Anyway, Any help, insight, or suggestion regarding this matter would be greatly appreciated. Also, I'm always happy to contribute on a pull request if requested.

Cheers !

**Comment by jts** [comment]
This looks like a bug to me (and running your example through valgrind). I'm not sure how to fix this without changing the interface to the `fetch_seq` function since it returns a slice. As a possible fix I removed the `fetch_seq` function and fixed the leak in `fetch_seq_string` in this commit: https://github.com/jts/rust-htslib/commit/6334e9a86e15d9a944851925fa8d09f3df478294. There might be a better way though.

**Comment by dlaehnemann** [comment]
I guess the `libc::free(ptr as *mut c_void);` is the decisive change, right here:
https://github.com/jts/rust-htslib/commit/6334e9a86e15d9a944851925fa8d09f3df478294#diff-dd23e4893230213542657967d1811dcd664d135f1bf0f41e1264875b193d086fR88

Wouldn't it be possible to keep the original setup and just do the free like this:
```
        let cseq = unsafe {
            let ptr = htslib::faidx_fetch_seq64(
                self.inner,                          //*const faidx_t,
                cname.as_ptr(),                      // c_name
                begin as htslib::hts_pos_t,          // p_beg_i
                end as htslib::hts_pos_t,            // p_end_i
                &mut (len_out as htslib::hts_pos_t), //len
            );
            s = ffi::CStr::from_ptr(ptr)
            libc::free(ptr as *mut c_void);
            s
        };

        Ok(cseq.to_bytes()
```

Also, maybe it makes sense to include the above MRE as a test case, to prevent regressions? But I'm not sure how big the test case is, and how easy it is to test for the memory leak... So it might be more important to get a fix in quickly...

**Comment by jts** [comment]
I don't think that version works because `ffi::CStr::from_ptr` doesn't make a copy of `ptr`'s data, it only wraps it. So it is no longer valid after the `libc::free` call. I played around with a few other ideas for keeping the interface the same, but they all required returning a slice referencing a local copy of the data, which the borrow checker didn't allow. That's not to say its impossible, I just wasn't able to make it work.

**Comment by dlaehnemann** [comment]
Then what about:

```
        let cseq = unsafe {
            let ptr = htslib::faidx_fetch_seq64(
                self.inner,                          //*const faidx_t,
                cname.as_ptr(),                      // c_name
                begin as htslib::hts_pos_t,          // p_beg_i
                end as htslib::hts_pos_t,            // p_end_i
                &mut (len_out as htslib::hts_pos_t), //len
            );
            s = ffi::CStr::from_ptr(ptr).to_bytes()
            libc::free(ptr as *mut c_void);
            s
        };

        Ok(cseq)
```

Does the `to_bytes()` clone the actual data?

**Comment by dlaehnemann** [comment]
And if this doesn't do the trick, maybe a `.clone()` can? It should be implemented on `CStr`, see:
https://doc.rust-lang.org/std/ffi/struct.CStr.html#impl-Clone-for-Box%3CCStr,+Global%3E

**Comment by jts** [comment]
I don't think to_bytes() clones. I just tested it (either calling to_bytes() in the return, or the second version that does it in the unsafe block) and valgrind complains about reading the memory block after free was called. 

This doesn't seem to work either:

```
[...]
            let s = ffi::CStr::from_ptr(ptr).to_bytes().clone();
            libc::free(ptr as *mut c_void);
            s
        };

        Ok(cseq)
    }
```

Maybe it is cloning the ptr's address, not the actual data

**Comment by MaelLefeuvre** [comment]
Hi everyone,

Thank you all for taking the matter at hand, and sorry for the late response (was at a conference).

I played around a bit with the source code and found that @dlaehnemann

**Comment by wdecoster** [comment]
Is there something here that I can do to get this fix implemented? I am seeing the same memory problem in my application.

**Comment by MaelLefeuvre** [comment]
> Is there something here that I can do to get this fix implemented? I am seeing the same memory problem in my application.

If you don't wish to use a forked project, a temporary workaround would simply be to manually call [`libc::free`](https://docs.rs/libc/0.2.147/libc/fn.free.html) on your fetched fasta sequence, once not needed.

A pseudo code example: 
```rust
loop {
    /* Setup code*/
    let refseq = reference.fetch_seq(chr, start, end)?;

    /* Do stuff with refseq */

    unsafe {libc::free(refseq.as_ptr() as *mut std::ffi::c_void)}; // Free up memory
}
```

You'll probably need to import [libc](https://docs.rs/libc/0.2.147/libc/) within your `Cargo.toml` file, i.e.:
```toml
[dependencies]
  libc = "0.2"
```

Hope this helps, until a more stable solution to the issue is found.

Cheers

**Comment by wdecoster** [comment]
Excellent, thank you for the detailed solution!
